packages/@ember/-internals/extension-support/lib/data_adapter.js

```diff
@@ -1,8 +1,119 @@
 import { getOwner } from '@ember/-internals/owner';
-import { scheduleOnce } from '@ember/runloop';
-import { get, objectAt, addArrayObserver, removeArrayObserver } from '@ember/-internals/metal';
+import { backburner } from '@ember/runloop';
+import { get } from '@ember/-internals/metal';
 import { dasherize } from '@ember/string';
 import { Namespace, Object as EmberObject, A as emberA } from '@ember/-internals/runtime';
+import { consumeTag, createCache, getValue, tagFor, untrack } from '@glimmer/validator';
+
+function iterate(arr, fn) {
+  if (Symbol.iterator in arr) {
+    for (let item of arr) {
+      fn(item);
+    }
+  } else {
+    arr.forEach(fn);
+  }
+}
+
+class RecordsWatcher {
+  recordCaches = new Map();
+
+  constructor(records, recordsAdded, recordsUpdated, recordsRemoved, wrapRecord, release) {
+    let { recordCaches } = this;
+
+    this.release = release;
+
+    let added, updated, removed;
```

Comment:
Why are these declared outside of the cache? It doesn't seem like they are used anywhere else, wouldn't it make more sense to declare them inside?

Reply:
I actually thought so too! But they are actually captured by the _inner_ caches, the ones created for each item. So, they have to be a shared reference.

Reply:
Hmm, I don't see how 🤔

Reply:
Specifically here: https://github.com/emberjs/ember.js/pull/19379/files#diff-f98ed9e6e2838cfeb4d67e382b98069cc174b21c1fec5b390faa83cdc9f7a43aR49

The cache for the item captures the reference to `updated`, and then it gets cached. Later on, it pushes into the same `updated` array the next time we load the cache. If we create a new `updated` array, then it's holding a reference to an older `updated`.

I can try to think about how to make this more intuitive, agree it's not super clear.

Reply:
Here's an alternative implementation, do you think this is clearer?

```js
class RecordsWatcher {
  recordCaches = new Map();

  added = [];
  updated = [];
  removed = [];

  getCacheForItem(record) {
    let recordCache = this.recordCaches.get(record);

    if (!recordCache) {
      let hasBeenAdded = false;

      recordCache = createCache(() => {
        if (!hasBeenAdded) {
          this.added.push(this.wrapRecord(record));
          hasBeenAdded = true;
        } else {
          this.updated.push(this.wrapRecord(record));
        }
      });

      this.recordCaches.set(record, recordCache);
    }
  }

  constructor(records, recordsAdded, recordsUpdated, recordsRemoved, wrapRecord, release) {
    this.release = release;
    this.wrapRecord = wrapRecord;

    this.recordArrayCache = createCache(() => {
      let seen = new Set();

      // Track `[]` for legacy support
      consumeTag(tagFor(records, '[]'));

      iterate(records, (record) => {
        getValue(this.getCacheForItem(record));
        seen.add(record);
      });

      // Untrack this operation because these records are being removed, they
      // should not be polled again in the future
      untrack(() => {
        this.recordCaches.forEach((cache, record) => {
          if (!seen.has(record)) {
            this.removed.push(wrapRecord(record));
            this.recordCaches.delete(record);
          }
        });
      });

      if (this.added.length > 0) {
        recordsAdded(this.added);
        this.added = [];
      }

      if (this.updated.length > 0) {
        recordsUpdated(this.updated);
        this.updated = [];
      }

      if (this.removed.length > 0) {
        recordsRemoved(this.removed);
        this.removed = [];
      }
    });
  }

  revalidate() {
    getValue(this.recordArrayCache);
  }
}
```